--- qemu/migration/ram.c	2017-08-11 14:53:14.847840561 +0800
+++ qemu-modified/migration/ram.c	2017-08-11 15:10:03.387820137 +0800
@@ -944,8 +944,15 @@
     int pages = 0;
     MemoryRegion *mr;
 
-    if (!block)
+    if (!block) {
         block = QLIST_FIRST_RCU(&ram_list.blocks);
+        /* In MicroCheckpointing, if last_seen_block is NULL and
+         * there is no drity block, we will loop infinitely along
+         * the ram_list.blocks list to find next dirty block
+         * because we can't identify that we have iterated all the blocks.
+         */
+        last_seen_block = block;
+    }
 
     while (true) {
         mr = block->mr;

--- qemu/migration/rdma.c	2017-08-11 14:53:14.847840561 +0800
+++ qemu-modified/migration/rdma.c	2017-08-11 15:11:18.975818607 +0800
@@ -1512,8 +1512,14 @@
 
         clear_bit(chunk, block->transit_bitmap);
 
-        if (rdma->nb_sent > 0) {
+        if (lc->nb_sent > 0) {
             rdma->nb_sent--;
+            lc->nb_sent--;
+            /*
+             * The source will freeze after calling ibv_get_cq_event() because
+             * there is no more WR in queue. This is caused by the improper
+             * maintainance of RDMALocalContext.nb_sent.
+             */
         }
 
         /*
